### PR TITLE
Fix setColumnOrder if 2 header rows

### PR DESCRIPTION
**Pull request by jcgueriaud1**
Change the HashMap to ConcurrentHashMap.
Change the test case to throw an exception if Map is a HashMap

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/828)

### CODE REVIEW

**Comment by jcgueriaud1** [comment]
Revert my change because it was wrong.
The GridColumnOrderTest was already broken in Java 11 or 12 so I didn't need to change this.

**Comment by tomivirkki** [comment]
If possible, please add a test for this use-case also (setColumnOrder with multiple header rows) to verify the bug and the fix for it.

**Comment by jcgueriaud1** [comment]
I was about to add a test but this test https://github.com/vaadin/vaadin-grid-flow/blob/master/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridColumnOrderTest.java#L78 is already there.

And it's not working (at least on my computer) in Java 11 and 12.


```
java.util.ConcurrentModificationException
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1134)
	at com.vaadin.flow.component.grid.GridColumnOrderHelper$GraphNodeLeafCache.getColumnIDs(GridColumnOrderHelper.java:292)
	at com.vaadin.flow.component.grid.GridColumnOrderHelper$GraphNodeLeafCache.findFirstContaining(GridColumnOrderHelper.java:312)
	at com.vaadin.flow.component.grid.GridColumnOrderHelper.reorderColumnsAndConsumeIDs(GridColumnOrderHelper.java:187)
	at com.vaadin.flow.component.grid.GridColumnOrderHelper.setColumnOrder(GridColumnOrderHelper.java:85)
	at com.vaadin.flow.component.grid.Grid.setColumnOrder(Grid.java:3825)
	at com.vaadin.flow.component.grid.Grid.setColumnOrder(Grid.java:3794)
	at com.vaadin.flow.component.grid.GridColumnOrderTest.setColumnOrder_simpleCaseWithHeader(GridColumnOrderTest.java:83)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

**Comment by petrixh** [comment]
___
*[vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/GridColumnOrderHelper.java, line 288 at r2](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/828#-Lu6jXoD2hkIiIJqIVS_:-Lu6mzxu8T581kTvi9d8:b7h3kgh) ([raw file](https://github.com/vaadin/vaadin-grid-flow/blob/921f4523085fa1f7f42088c10ee77afeb0b43906/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/GridColumnOrderHelper.java#L288)):*
<details><summary><i>Previously, petrixh (Petri Heinonen (Vaadin) ) wrote…</i></summary><blockquote>

The contract states that if the input column is a Grid.Column the returned set is a singleton. Is this still true? (thinking about the new empty HashSet() that is being returned which is more of an empty set rather than a singleton set, but not sure which branch it ends up in in the case of Grid.Column, then again it's a private method which I noticed only after writing this...). 
</blockquote></details>

Basically my thinking is, if this is still true, great, otherwise we should consider removing the second line in the contract as the next dev might rely on this being so.

**Comment by jcgueriaud1** [comment]
I updated the test class to add a test case with a lot of header rows and multi/single select case.
@tomivirkki Do you want a test case with TestBench ?

**Comment by tomivirkki** [comment]
Probably a TestBench test isn't necessary if this can be tested without it

**Comment by jcgueriaud1** [comment]
Did the test on master run on Java 9+ ?
Because I don't understand how it can run without errors.

**Comment by jcgueriaud1** [comment]
Add a test case with multiple header rows on a Treegrid as setColumnOrder was not tested automatically on a treegrid.
